schema/json/internal/metadata_types.go

```diff
@@ -0,0 +1,150 @@
+package internal
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/scylladb/go-set/strset"
+)
+
+var metadataExceptions = strset.New(
+	"FileMetadata",
+)
+
+func AllSyftMetadataTypeNames() ([]string, error) {
+	root, err := repoRoot()
+	if err != nil {
+		return nil, err
+	}
+	files, err := filepath.Glob(filepath.Join(root, "syft/pkg/*.go"))
+	if err != nil {
+		return nil, err
+	}
+	return findMetadataDefinitionNames(files...)
+}
```

Comment:
There is some related data [here](https://github.com/anchore/syft/blob/main/syft/pkg/metadata.go#L46), could you use any of these exported lists? These need to be kept up to date, used when decoding CDX.

Reply:
I thought about using these, but the `pkg.MetadataType` will be something that is removed closer to syft 1.0 https://github.com/anchore/syft/issues/1735 which means this map would be a CycloneDX-only format concern. So instead I leaned towards the current ast approach, which has the added benefit of not needing to remember to add it to a central list/map, in which case the PR protection would be a little more brittle.

Reply:
Right, but the list of all metadata types is something needed by the _Syft format decoder_ once the pkg.MetadataType goes away, which is specifically a concern with the Syft schema, is it not?

Reply:
I see -- that is, we should have a single source of truth for these sets. I'll try to account for that in this PR (warning, the scope will creep a bit, but I think it's warranted).

Reply:
I still agree with trying to get down to a single source of truth, however, I feel that is better tackled in https://github.com/anchore/syft/issues/1735 and https://github.com/anchore/syft/issues/1844 since there are naming inconsistencies and decoding considerations to make here. Additionally, this PR doesn't add another source of truth, this container struct being created exists today and is now being generated. Hopefully with the other two issues they can expand on this generation to include more things.